widgetssdk/src/main/java/com/glia/widgets/chat/ChatView.java

```diff
@@ -1073,15 +1135,121 @@ private static Uri chooseUriByRequestCode(int requestCode, Uri galeryImgUri, Uri
     public void onActivityResult(int requestCode, int resultCode, Intent intent) {
         if ((requestCode == OPEN_DOCUMENT_ACTION_REQUEST || requestCode == CAPTURE_IMAGE_ACTION_REQUEST || requestCode == CAPTURE_VIDEO_ACTION_REQUEST)
                 && resultCode == Activity.RESULT_OK) {
-
-            Uri uri = chooseUriByRequestCode(requestCode, intent.getData(), photoCaptureFileUri);
-
+            Uri dataUri = intent != null ? intent.getData() : null;
+            Uri uri = chooseUriByRequestCode(requestCode, dataUri, photoCaptureFileUri);
             if (uri != null) {
                 controller.onAttachmentReceived(uri);
             }
         }
     }
 
+    @Override
+    public void onFileDownloadClick(OperatorAttachmentItem clickedItem) {
+        submitUpdatedItems(clickedItem, true, false);
+
+        Context context = this.getContext();
+        AttachmentFile clickedFile = clickedItem.attachmentFile;
+        File file = new File(context.getFilesDir(), clickedFile.getName());
+        downloadFile(clickedItem, file);
+    }
+
+    private void downloadFile(OperatorAttachmentItem clickedItem, File file) {
+        if (clickedItem.attachmentFile.isDeleted()) {
+            Toast.makeText(this.getContext(), this.getContext().getString(R.string.chat_file_download_failed_msg), Toast.LENGTH_SHORT).show();
+            return;
+        }
+
+        Glia.fetchFile(clickedItem.attachmentFile, (fileInputStream, gliaException) -> new Thread(() -> {
+            try {
+                try (OutputStream output = new FileOutputStream(file)) {
```

Comment:
extract this to utils

Reply:
I think it does not change much. This call should not be here at all but the current implementation is full of such calls.

Reply:
would make the chatview bit shorter at least - this is sort of file handling methods anyway

Reply:
Alright, will move it then.